don't use builtins.fetchGit for cargo git dependencies

sysedwinistrator opened the pull request:
## Summary

The goal of this PR is to remove the usage of `builtins.fetchGit` during evaluation (see below for the reasons).

To achieve this, it:

- Replaces `cargoLock` with `cargoHash`
  - `cargoHash` needs to be updated each time the upstream lockfile changes
  - simply updating the Flake inputs therefore won't work anymore
  - [nix-update](https://github.com/Mic92/nix-update) can do this automatically when updating packages [^1]
- Refactors the packages into expressions that:
  - use `fetchFromGitHub` instead of flake inputs as `src`
  - can be updated automatically by nix-update
  - be called via `callPackage` independently of the rest of this flake (was on @sodiboo's wishlist)
- Removes the flake inputs for package sources
- Extends the CI workflow to run nix-update for the unstable pacakges
  - Updating the flake inputs is still done for nixpkgs
- Makes necessary minor adjustments to the changes and removes dead code

[^1]: nix-update could also be used to _only_ update the cargoHash.
---

## Background

### Context
Yesterday, when I was making changes to my config, `home-manager switch` got stuck because https://gitlab.freedesktop.org/pipewire/pipewire-rs was unreachable (apparently it was being (D)DoS'ed yesterday).  
This struck me as odd, because I had already built the niri-unstable pkg (and any other packages used by my configuration, I only changed some keybindings). Usually the source repo only gets fetched when building a package.  
It turns out this is happening because the flake allows `buildRustPackage` to use `builtins.fetchGit`, which fetches during evaluation.

I also saw some issues in this repo related to this: https://github.com/sodiboo/niri-flake/issues/990 https://github.com/sodiboo/niri-flake/issues/923.

Please note that https://gitlab.freedesktop.org is run by volunteers, doesn't have infinite resources, and has had outages from (D)DoS more than once.  
So it would be both in the interest of the users of this flake (by being able to use it even when https://gitlab.freedesktop.org is down) and the operators of https://gitlab.freedesktop.org (by not being subjected to unnecessary git pulls) to switch to nixpkgs fetchers.

The downside of this is that the output hashes for all git dependencies have to be specified.  
They will also need to be updated if upstream updates or adds git dependencies. In that case, the update automation this repo uses might break.

### Explanation (original commit message, outdated)
Unlike nixpkgs fetchers (such as pkgs.fetchgit or pkgs.fetchFromGitLab), which are fixed-output derivations and therefore do the actual fetching only when they're being built; builtins.fetchGit fetches when the expression is being evaluated (regardless of whether the package for which it provides the source is even being built). Also, because it doesn't evaluate to a derivation that can be built, its result cannot be pushed to a binary cache.

This means that the git repos using builtins.fetchGit need to be fetched each time a package using them is being evaluated (although there's some level of local caching), even when the package itself has already been built.

In contrast, when using nixpkgs fetchers for src, the src attribute is basically just another derivation the package depends on, and will be built (or substituted!) when building the package.

It's for these reasons that builtin fetchers such as builtins.fetchGit aren't being used in nixpkgs.

sodiboo commented:
I am aware of this and want to fix it; indeed i should be specifying fixed hashes. But simply adding the hashes is not a complete solution because i also have automations to update the input sources in this repo automatically. The existing workflows are only based around updating flake inputs; and this is the entire reason why i use builtin fetchGit (it's simply easier). To use fixed hashes, the workflows need to update those hashes, and it needs to be specified for all four packages (`{niri,xwayland-satellite}-{stable,unstable}`) and updated individually.  At that point, it might also be worth it to no longer fetch the niri sources with flake inputs, and instead update them too using the same nixpkgs update scripts. In the same move, i'd probably want to separate the `{stable,unstable}` pairs into separate functions entirely, and probably just put them all in separate files that can be passed to `pkgs.callPackage` without necessarily consuming this repo as a flake.

So, if you feel like doing so, and, *shudders*, want to deal with GitHub Actions, be my guest and update this PR. If you don't feel like doing so, you can close this PR, and i pinky promise that i will eventually get around to it.

sysedwinistrator commented:
I'm trying out https://github.com/Mic92/nix-update for this.   

I got ahead of myself and removed the lockfile import that triggers IFD (`lockFile = "${src}/Cargo.lock";`). But apparently nix-update cannot update the vendored Cargo.lock, it can only generate a completely new lock file (that is out of sync with upstrem).

sysedwinistrator commented:
I modified the workflow to run nix-update before updating flake.lock. It's a bit rudimentary, as it's using the update-flake-lock step after it to commit the changes. Also, the job is still named `lockfile`.

Here's a run of the workflow in my fork: https://github.com/sysedwinistrator/niri-flake/actions/runs/17805089719/job/50614809720. It successfully created a PR that updates both flake.lock and niri-unstable: https://github.com/sysedwinistrator/niri-flake/pull/1/files. The rest of the `update.yml` workflow failed because I don't have the secrets setup in my fork.

sodiboo commented:
As far as i can tell, the `cargoHash` attribute is perhaps simpler than the lockfile IFD? This is a hash that, i suppose, is the result of reading the lockfile from `src` and fetching all the dependencies in one go, rather than fetching each dependency individually. So, it doesn't require keeping a lockfile nearby (and also doesn't require IFD).  `nix-update` says it supports updating `cargoHash`. this is the method [currently used in nixpkgs](https://github.com/NixOS/nixpkgs/blob/6e6ad7b21cf6cda90fa7bac7397a284b2ba232c4/pkgs/by-name/ni/niri/package.nix#L50).

sodiboo commented on .github/workflows/update.yml:
```diff
@@ -18,12 +18,18 @@ jobs:
-      - name: Update flake.lock
+      - name: Update versions
+        run: |
+          nix run nixpkgs#nix-update -- --flake --version=stable --override-file=pkgs/niri/stable/default.nix niri-stable
+          nix run nixpkgs#nix-update -- --flake --version=branch --override-file=pkgs/niri/unstable/default.nix niri-unstable
+          nix run nixpkgs#nix-update -- --flake --version=stable --override-file=pkgs/xwayland-satellite/stable/default.nix xwayland-satellite-stable
+          nix run nixpkgs#nix-update -- --flake --version=branch --override-file=pkgs/xwayland-satellite/unstable/default.nix xwayland-satellite-unstable
```
nix-update should *not* be automatically updating the stable packages here

or at least, not for niri. maybe for xwayland-satellite? haven't thought too much about this.

but, generally, i ensure that the settings module always exhaustively supports `niri-stable`. it must therefore only be manually updated.

it would be nice if the xwayland-satellite package was automatically updated but not automatically merged? so i get a nixpkgs-like workflow where the bot pings me "hey, this can be updated pretty please". niri could behave like this too. but this should not be in the automatically-merged workflow.

ultimately, that's for me to figure out how exactly i want to deal with it. since i'm the one who'll be manually merging the stable updates. so for now, just remove the lines updating the stable packages.

```suggestion
          nix run nixpkgs#nix-update -- --flake --version=branch --override-file=pkgs/niri/unstable/default.nix niri-unstable
          nix run nixpkgs#nix-update -- --flake --version=branch --override-file=pkgs/xwayland-satellite/unstable/default.nix xwayland-satellite-unstable
```

sysedwinistrator committed:
refactor to conventional package structure for nix-update

- Remove usage of builtins.fetchGit by using cargoHash instead of
  cargoLock (nix-update can update cargoHash automatically).
- Refactor packages to conventional (nixpkgs-like) file structure
  compatible with nix-update.
- Get src using fixed-output-derivation (fetchFromGitHub) instead of
  source-only flake inputs. This is a requirement for nix-update.
- Set unstable versions in format recognized by nix-update.
- Remove logic for fetching and looking up versions by their Git rev.
- Use callPackage to a generic.nix in a way that keeps the final
  derivation's arguments overridable (they won't be discoverable via
  <pkg>.override.__functionArgs, though).

sysedwinistrator committed:
run nix-update against unstable pkgs in GitHub workflow

sysedwinistrator committed:
add target to update stable pkgs to justfile

sysedwinistrator committed:
fix references to removed flake inputs
